core/src/main/java/org/apache/iceberg/TableUtil.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class TableUtil {
+  private TableUtil() {}
+
+  /**
+   * Returns the format version of the given table. Note that subclasses of {@link
+   * BaseMetadataTable} do not have a format version. The only exception to this is {@link
+   * PositionDeletesTable}, where the format version is fetched from the underlying table.
+   */
+  public static int formatVersion(Table table) {
```

Comment:
I agree with this approach. I just wonder if we should not define the format version just as table property (and set when using delete vectors for instance). It's more "hardcoded" but probably easier to evaluate (and fallback to V2 if the table property is not present).

Reply:
I would strongly prefer we keep it out of table properties. IMHO, Table properties are user facing hints or guides and the system should not rely on them for correct behavior. Having the correct format version is important for correct behavior so I would keep it out of properties.